gtsam/slam/EssentialMatrixFactor.h

```diff
@@ -267,28 +269,128 @@ class EssentialMatrixFactor3: public EssentialMatrixFactor2 {
+ * prior should have a noise model with very low sigma in the remaining
+ * dimensions. This has been tested to work on Cal3_S2. With Cal3Bundler, it
+ * works only with a strong prior (low sigma noisemodel) on all degrees of
+ * freedom.
+ */
+template <class CALIBRATION>
+class EssentialMatrixFactor4
+    : public NoiseModelFactor2<EssentialMatrix, CALIBRATION> {
+ private:
+  Point2 pA_, pB_;  ///< points in pixel coordinates
+
+  typedef NoiseModelFactor2<EssentialMatrix, CALIBRATION> Base;
+  typedef EssentialMatrixFactor4 This;
+
+  static const int DimK = FixedDimension<CALIBRATION>::value;
+  typedef Eigen::Matrix<double, 2, DimK> JacobianCalibration;
+
+ public:
+  /**
+   *  Constructor
+   *  @param keyE Essential Matrix (from camera B to A) variable key
+   *  @param keyK Calibration variable key (common for both cameras)
+   *  @param pA point in first camera, in pixel coordinates
+   *  @param pB point in second camera, in pixel coordinates
+   *  @param model noise model is about dot product in ideal, homogeneous
+   * coordinates
+   */
+  EssentialMatrixFactor4(Key keyE, Key keyK, const Point2& pA, const Point2& pB,
+                         const SharedNoiseModel& model)
+      : Base(model, keyE, keyK), pA_(pA), pB_(pB) {}
+
+  /// @return a deep copy of this factor
+  gtsam::NonlinearFactor::shared_ptr clone() const override {
+    return boost::static_pointer_cast<gtsam::NonlinearFactor>(
+        gtsam::NonlinearFactor::shared_ptr(new This(*this)));
+  }
+
+  /// print
+  void print(
+      const std::string& s = "",
+      const KeyFormatter& keyFormatter = DefaultKeyFormatter) const override {
+    Base::print(s);
+    std::cout << "  EssentialMatrixFactor4 with measurements\n  ("
+              << pA_.transpose() << ")' and (" << pB_.transpose() << ")'"
+              << std::endl;
+  }
+
+  /**
+   * @brief Calculate the algebraic epipolar error pA' (K^-1)' E K pB.
+   *
+   * @param E essential matrix for key keyE
+   * @param K calibration (common for both images) for key keyK
+   * @param H1 optional jacobian of error w.r.t E
+   * @param H2 optional jacobian of error w.r.t K
+   * @return * Vector 1D vector of algebraic error
+   */
+  Vector evaluateError(
+      const EssentialMatrix& E, const CALIBRATION& K,
+      boost::optional<Matrix&> H1 = boost::none,
+      boost::optional<Matrix&> H2 = boost::none) const override {
+    // converting from pixel coordinates to normalized coordinates cA and cB
+    JacobianCalibration cA_H_K;  // dcA/dK
+    JacobianCalibration cB_H_K;  // dcB/dK
+    Point2 cA = K.calibrate(pA_, H2 ? &cA_H_K : 0, boost::none);
+    Point2 cB = K.calibrate(pB_, H2 ? &cB_H_K : 0, boost::none);
+
+    // convert to homogeneous coordinates
+    Vector3 vA = EssentialMatrix::Homogeneous(cA);
+    Vector3 vB = EssentialMatrix::Homogeneous(cB);
+
+    if (H2) {
+      // compute the jacobian of error w.r.t K
+
+      // error function f = vA.T * E * vB
```

Comment:
do we define `vA` anywhere?

Reply:
In lines above: 
```
    // convert to homogeneous coordinates   
     Vector3 vA = EssentialMatrix::Homogeneous(cA);
```
Is it not as obvious as I thought? We can add documentation then.